manticore/native/state.py

```diff
@@ -25,28 +58,36 @@ def execute(self):
             ConcretizeRegister,
         )  # must be here, otherwise we get circular imports
 
+        self._checkpoint()
         try:
             result = self._platform.execute()
 
         # Instead of State importing SymbolicRegisterException and SymbolicMemoryException
         # from cpu/memory shouldn't we import Concretize from linux, cpu, memory ??
         # We are forcing State to have abstractcpu
-        except ConcretizeRegister as e:
+        except ConcretizeRegister as exc:
+            # Need to define local variable to use in closure
+            e = exc
             expression = self.cpu.read_register(e.reg_name)
 
-            def setstate(state, value):
-                state.cpu.write_register(setstate.e.reg_name, value)
+            def setstate(state: State, value):
+                state.cpu.write_register(e.reg_name, value)
 
-            setstate.e = e
+            self.rollback(self._checkpoint_data)
             raise Concretize(str(e), expression=expression, setstate=setstate, policy=e.policy)
-        except ConcretizeMemory as e:
+        except ConcretizeMemory as exc:
+            # Need to define local variable to use in closure
+            e = exc
             expression = self.cpu.read_int(e.address, e.size)
 
-            def setstate(state, value):
-                state.cpu.write_int(setstate.e.address, value, setstate.e.size)
+            def setstate(state: State, value):
+                state.cpu.write_int(e.address, value, e.size)
```

Comment:
> What is going on here!? In particular, does that e = exc assignment actually do anything?!

I think this is weird due to the contextmanager for the exception. I can't explain it, but when I remove it, it fails with an "unbounded variable" (I forget exactly)

> Additionally, it looks like the checkpoint & rollback logic is only used within this single method. If that's the case, does the checkpoint data actually need to ever be stored within the state at all? Or would it be possible to simply use local variables within the execute() method, and avoid having to add to the state at all?

That's a fair point. I was basing my solution off of what was done in EVM, but since we don't have anything near as feature-filled in native, we can avoid any unnecessary complexity surrounding this until it's needed.

Reply:
> > What is going on here!? In particular, does that e = exc assignment actually do anything?!
>
> I think this is weird due to the contextmanager for the exception. I can't explain it, but when I remove it, it fails with an "unbounded variable" (I forget exactly)

Oh! I remember what's happening here now!

**In Python 3, a variable that is bound by an exception handler is implicitly unbound after exception-handling.**  This is done to avoid hanging on to local variables in the exception object for longer than needed.  This is a Python 3 change, and was not the behavior in Python 2.  Read more about it here if you're interested: https://docs.python.org/3/reference/compound_stmts.html#the-try-statement

I bet that this exact behavior was what was causing trouble.  A small test case:
```python
def test(should_raise):
    try:
        if should_raise:
            raise ValueError(42)
    except ValueError as e:
        def callback(prefix):
            print('%s: got an exception: %s' % (prefix, e))
        return callback

    def callback(prefix):
        print('%s: completed successfully!' % prefix)
    return callback
```
Then, in the non-throwing case, it works as expected:
```
>>> test(should_raise=False)('case 1')
case 1: completed successfully!
```
But in the throwing case, you get trouble:
```
>>> test(should_raise=True)('case 2')
Traceback (most recent call last):
  File "test.py", line 16, in <module>
    test(should_raise=True)('case 2')
  File "test.py", line 7, in callback
    print('%s: got an exception: %s' % (prefix, e))
NameError: free variable 'e' referenced before assignment in enclosing scope
```
Note, if you run these examples in Python 2, they both work without trouble; it's only in Python 3 that the behavior changed.  Additionally, if you change the `test()` function slightly, so that the exception handler looks like this instead (as done here in this PR), it works fine:
```python
    except ValueError as e:
        exn = e  # explicitly re-bind exception variable
        def callback(prefix):
            print('%s: got an exception: %s' % (prefix, exn))
        return callback
```